lib_standard_app/debug.h

```diff
@@ -1,3 +1,168 @@
+#define LR_AND_PC_OFFSET 0
+#define LR_AND_PC_SIZE   30
+WEAK void assert_display_lr_and_pc(int lr, int pc);
+#define ASSERT_DISPLAY_LR_AND_PC(lr, pc) assert_display_lr_and_pc(lr, pc)
+#else
+#define LR_AND_PC_SIZE 0
+#endif
+
+#ifdef LEDGER_ASSERT_CONFIG_MESSAGE_INFO
+#define MESSAGE_OFFSET LR_AND_PC_SIZE
+#define MESSAGE_SIZE   20
+WEAK void assert_display_message(const char *message);
+#define ASSERT_DISPLAY_MESSAGE(message) assert_display_message(message)
+#else
+#define MESSAGE_SIZE 0
+#endif
+
+#ifdef LEDGER_ASSERT_CONFIG_FILE_INFO
+#define FILE_OFFSET MESSAGE_OFFSET + MESSAGE_SIZE
+#define FILE_SIZE   50
+WEAK void assert_display_file_info(const char *file, unsigned int line);
+#define ASSERT_DISPLAY_FILE_INFO(file, line) assert_display_file_info(file, line)
+#else
+#define FILE_SIZE 0
+#endif
+
+#define ASSERT_BUFFER_LEN LR_AND_PC_SIZE + MESSAGE_SIZE + FILE_SIZE
+#else
+#ifdef LEDGER_ASSERT_CONFIG_LR_AND_PC_INFO
+#define ASSERT_DISPLAY_LR_AND_PC(lr, pc) \
+    do {                                 \
+    } while (0)
+#endif
+
+#ifdef LEDGER_ASSERT_CONFIG_MESSAGE_INFO
+#define ASSERT_DISPLAY_MESSAGE(message) \
+    do {                                \
+    } while (0)
+#endif
+
+#ifdef LEDGER_ASSERT_CONFIG_FILE_INFO
+#define ASSERT_DISPLAY_FILE_INFO(file, line) \
+    do {                                     \
+    } while (0)
+#endif
+#endif
+
+#ifdef HAVE_PRINTF
+#ifdef LEDGER_ASSERT_CONFIG_LR_AND_PC_INFO
+void print_lr_and_pc(int lr, int pc);
+#define ASSERT_PRINT_LR_AND_PC(lr, pc) print_lr_and_pc(lr, pc)
+#endif
+
+#ifdef LEDGER_ASSERT_CONFIG_MESSAGE_INFO
+void print_message(const char *message);
+#define ASSERT_PRINT_MESSAGE(message) print_message(message)
+#endif
+
+#ifdef LEDGER_ASSERT_CONFIG_FILE_INFO
+void print_file_info(const char *file, int line);
+#define ASSERT_PRINT_FILE_INFO(file, line) print_file_info(file, line)
+#endif
+#else
+#ifdef LEDGER_ASSERT_CONFIG_LR_AND_PC_INFO
+#define ASSERT_PRINT_LR_AND_PC(lr, pc) \
+    do {                               \
+    } while (0)
+#endif
+
+#ifdef LEDGER_ASSERT_CONFIG_MESSAGE_INFO
+#define ASSERT_PRINT_MESSAGE(message) \
+    do {                              \
+    } while (0)
+#endif
+
+#ifdef LEDGER_ASSERT_CONFIG_FILE_INFO
+#define ASSERT_PRINT_FILE_INFO(file, line) \
+    do {                                   \
+    } while (0)
+#endif
+#endif
+
+#ifdef LEDGER_ASSERT_CONFIG_LR_AND_PC_INFO
+#define LEDGER_ASSERT_LR_AND_PC()                            \
+    do {                                                     \
+        int _lr_address = 0;                                 \
+        int _pc_address = 0;                                 \
+                                                             \
+        __asm volatile("mov %0, lr" : "=r"(_lr_address));    \
+        __asm volatile("mov %0, pc" : "=r"(_pc_address));    \
+        _lr_address = compute_address_location(_lr_address); \
+        _pc_address = compute_address_location(_pc_address); \
```

Comment:
To save memory on apps, and as this will probably be the default mode, we could move the `compute_address_location` inside `print_lr_and_pc` and `assert_display_lr_and_pc`.

Reply:
very good point !